src/picongpu/include/particles/ionization/ionization.hpp

```diff
@@ -0,0 +1,386 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+
+#include <iostream>
+
+#include "simulation_defines.hpp"
+#include "particles/Particles.hpp" 
+#include <cassert>
+
+#include "mappings/kernel/AreaMapping.hpp"
+
+#include "particles/ParticlesInit.kernel"
+#include "mappings/simulation/GridController.hpp"
+
+#include "simulationControl/MovingWindow.hpp"
+
+#include <limits>
+
+#include "traits/Resolve.hpp"
+
+//#include "particles/traits/GetIonizer.hpp"
+#include "types.h"
+
+#include "particles/ionization/ionizationMethods.hpp"
+
+namespace picongpu
+{
+    
+using namespace PMacc;
+
+/* IONIZE PER FRAME */
+
+template<class IonizeAlgo, class TVec, class T_Field2ParticleInterpolation, class NumericalCellType>
+struct IonizeParticlesPerFrame
+{
+
+    template<class FrameType, class BoxB, class BoxE >
+    DINLINE void operator()(FrameType& ionFrame, int localIdx, BoxB& bBox, BoxE& eBox, unsigned int& newElectrons)
+    {
+
+        typedef TVec Block;
+        /* @new type for field to particle interpolation */
+        typedef T_Field2ParticleInterpolation Field2ParticleInterpolation;
+
+        typedef typename BoxB::ValueType BType;
+        typedef typename BoxE::ValueType EType;
+
+        PMACC_AUTO(particle,ionFrame[localIdx]);
+        const float_X weighting = particle[weighting_];
+        
+        floatD_X pos = particle[position_];
+        const int particleCellIdx = particle[localCellIdx_];
+
+        DataSpace<TVec::dim> localCell(DataSpaceOperations<TVec::dim>::template map<TVec > (particleCellIdx));
+
+       
+        EType eField = Field2ParticleInterpolation()
+            (eBox.shift(localCell).toCursor(), pos, NumericalCellType::getEFieldPosition());
+        BType bField =Field2ParticleInterpolation()
+            (bBox.shift(localCell).toCursor(), pos, NumericalCellType::getBFieldPosition());
+
+        float3_X mom = particle[momentum_];
+        const float_X mass = attribute::getMass(weighting,particle);
+        
+        /* define charge state before ionization */
+        float_X prevChState = particle[chargeState_];
+ 
+        /* this is the point where actual ionization takes place */
+        IonizeAlgo ionizeAlgo;
+        ionizeAlgo(
+             bField, eField,
+             particle
+             );
+        
+        particle[momentum_] = mom;
+        particle[position_] = pos;
+        /* determine number of new electrons to be created */
+        newElectrons = particle[chargeState_] - prevChState;
```

Comment:
this is a very sensitive piece of code for new users and it mixes already data structures ("particles"), algorithms and the "physical model" (like BSI). For that reason, a new reader shall be able to see immediately if we handle single or more "real" electrons at this point.